File: string-easy/28.实现-str-str.py
#
# @lc app=leetcode.cn id=28 lang=python3
#
# [28] 实现 strStr()
#

# @lc code=start
class Solution:
    # 78/78 cases passed (56 ms)
    # Your runtime beats 19.28 % of python3 submissions
    # Your memory usage beats 33.95 % of python3 submissions (15 MB)
    def strStr(self, haystack: str, needle: str) -> int:
        if len(needle) == 0:
            return 0
        ptr = 0
        i = 0
        while i < len(haystack):
            if haystack[i] == needle[ptr]:
                ptr += 1
            else:
                i -= ptr
                ptr = 0
            if ptr == len(needle):
                return i - ptr + 1
            i += 1
        return -1


# @lc code=end


class Solution_1:
    def strStr(self, haystack: str, needle: str) -> int:
        h, n = len(haystack), len(needle)
        for start in range(h - n + 1):
            if haystack[start : start + n] == needle:
                return start
        return -1

File: string-easy/28.实现-str-str.cpp
/*
 * @lc app=leetcode.cn id=28 lang=cpp
 *
 * [28] 实现 strStr()
 */

#include <string>

using namespace std;

// @lc code=start

// 79/79 cases passed (0 ms)
// Your runtime beats 100 % of cpp submissions
// Your memory usage beats 80.94 % of cpp submissions (6.7 MB)
class Solution {
   public:
    // 暴力匹配，时间复杂度 O(MN)
    int strStr(string txt, string pat) {
        int M = pat.length();
        int N = txt.length();
        for (int i = 0; i <= N - M; i++) {
            int j = 0;
            for (; j < M; j++) {
                if (pat[j] != txt[i + j])
                    break;
            }
            // pat 全都匹配了
            if (j == M)
                return i;
        }
        // txt 中不存在 pat 子串
        return -1;
    }
};
// @lc code=end
